fuse_loss/include/fuse_loss/qwt_loss_plot.h

```diff
@@ -0,0 +1,333 @@
+                   [&loss_function](const auto& r) {  // NOLINT(whitespace/braces)
+                     double rho[3];
+                     loss_function->Evaluate(r * r, rho);
+                     return r * rho[1];
+                   });  // NOLINT(whitespace/braces)
+
+    return influence;
+  }
+
+  std::vector<double> weight(const ceres::LossFunction* loss_function) const
+  {
+    std::vector<double> weight;
+    weight.reserve(residuals_.size());
+
+    // The weight is defined as the influence divided by the residual r:
+    //
+    //   w(r) = \phi(r) / r
+    //
+    // Since the loss functions we use are like Ceres solver ones, we have:
+    //
+    //              d rho(s)       d rho(s)
+    //   w(r) = r * -------- / r = --------
+    //                 ds             ds
+    //
+    // That is, rho[1].
+    std::transform(residuals_.begin(), residuals_.end(), std::back_inserter(weight),
+                   [&loss_function](const auto& r) {  // NOLINT(whitespace/braces)
+                     double rho[3];
+                     loss_function->Evaluate(r * r, rho);
+                     return rho[1];
+                   });  // NOLINT(whitespace/braces)
+
+    return weight;
+  }
+
+  std::vector<double> secondDerivative(const ceres::LossFunction* loss_function) const
+  {
+    std::vector<double> second_derivative;
+    second_derivative.reserve(residuals_.size());
+
+    // The second derivative is defined as:
+    //
+    //   d^2 rho(s)
+    //   ----------
+    //      ds^2
+    //
+    // That is, rho[2].
+    std::transform(residuals_.begin(), residuals_.end(), std::back_inserter(second_derivative),
+                   [&loss_function](const auto& r) {  // NOLINT(whitespace/braces)
+                     double rho[3];
+                     loss_function->Evaluate(r * r, rho);
+                     return rho[2];
+                   });  // NOLINT(whitespace/braces)
+
+    return second_derivative;
+  }
+
+  const std::vector<double>& getResiduals() const
+  {
+    return residuals_;
+  }
+
+private:
+  std::vector<double> residuals_;
+};
+
+class QwtLossPlot
+{
+public:
+  QwtLossPlot(const std::vector<double>& residuals, const HSVColormap& colormap)
+    : residuals_(QVector<double>::fromStdVector(residuals))
+    , loss_evaluator_(residuals)
+    , colormap_(colormap)
+    , magnifier_(plot_.canvas())
+    , zoomer_(plot_.canvas())
+    , panner_(plot_.canvas())
+  {
+    // Allocate one curve per colormap entry:
+    curves_.reserve(colormap_.size());
+
+    // Set background to white:
+    plot_.setCanvasBackground(Qt::white);
+
+    // Add grid:
+    grid_.setPen(Qt::lightGray, 0.0, Qt::DotLine);
+    grid_.attach(&plot_);
+
+    // Setup panner modifiers:
+    panner_.setMouseButton(Qt::LeftButton, Qt::ControlModifier);
+
+    // Add a legend:
+    plot_.insertLegend(&legend_);
+  }
+
+  void plotRho(const std::shared_ptr<fuse_core::Loss>& loss)
```

Comment:
I also was lazy here. These methods are almost the same, only changing the method called from the `LossEvaluator` class above.

Again, I'm not completely sure of the best way to organize this, and since it's a unit test, I didn't care much at the beginning. But maybe now that there are 4 methods, it's worth trying to factor things out. :thinking:

Reply:
It should be pretty trivial to factor out a `QwtPlotCurve* createCurve(name, values)` method. And maybe a `std::string getName(type)` so you don't have the `11` and `-15` in multiple places. But however you want to skin that cat is fine.

Reply:
Done.

Although I guess I could factor things out a bit more.